### PR TITLE
Metal: throw an NSException when attempting to draw with invalid program

### PR DESCRIPTION
**Pull request by bejado**
Throw an NSException when a program fails to compile and then is used for drawing; this helps aid debugging compiler errors in production, where stdout logs are not available.

b/325321035

### CODE REVIEW

File: filament/backend/src/metal/MetalShaderCompiler.h [hunk 1]
```diff
@@ -41,36 +43,74 @@ class MetalShaderCompiler {
...
+            NSString* reason =
+                    [NSString stringWithFormat:
+                            @"Attempting to draw with an id<MTLFunction> that "
+                            @"failed to compile. Compiler error: %@", errorMessage];
+            NSException* compilationFailureException =
+                    [NSException exceptionWithName:@"MetalCompilationFailure"
+                                            reason:reason
+                                          userInfo:nil];
+            [compilationFailureException raise];
+        }
```
Comment (pixelflinger) [inline filament/backend/src/metal/MetalShaderCompiler.h:74]: The "slow" path of vlaidate() should be in the cpp file, so it's not inlined. It should probably be marked as UTILS_NO_INLINE too.

**Review by pixelflinger — CHANGES_REQUESTED** [review]
I think this deserves a better description in the checking comment. Also is this closing a specific bug? (does it need an entry in readme or a FIXES tag?).

File: filament/backend/src/metal/MetalShaderCompiler.mm [hunk 2]
```diff
@@ -225,4 +229,22 @@ bool isReady() const noexcept {
     mCallbackManager.setCallback(handler, callback, user);
 }
 
+UTILS_NOINLINE
+void MetalShaderCompiler::MetalFunctionBundle::validate() const {
+    if (UTILS_LIKELY(!std::holds_alternative<Error>(mPrograms))) {
```
Comment (poweifeng) [inline filament/backend/src/metal/MetalShaderCompiler.mm:234]: couldn't this just make use of `isValid`?  Should `None` also throw something?

Comment (bejado) [reply filament/backend/src/metal/MetalShaderCompiler.mm]: Good point about `None`, I updated the code to also throw an exception when it encounters None.

**Review by poweifeng — APPROVED** [review]

**Review by pixelflinger — APPROVED** [review]

**Pull request merged**